java/oauth-custom-challengehandler/src/main/java/com/esri/arcgisruntime/sample/oauthcustomchallengehandler/MainActivity.java

```diff
@@ -0,0 +1,304 @@
+            WebSettings webSettings = webView.getSettings();
+            webSettings.setBuiltInZoomControls(true);
+            webSettings.setJavaScriptEnabled(true);
+
+            // setup webviewClient
+            CustomWebViewClient customWebViewClient = new CustomWebViewClient(portalUrl,
+                mOAuthConfiguration.getClientId(), mOAuthConfiguration.getRedirectUri(), oAuthChallengeCountDownLatch);
+            webView.setWebViewClient(customWebViewClient);
+
+            webView.loadUrl(url);
+            // set the webview on the dialog
+            mOAuthLoginDialog.setContentView(webView);
+            // display the dialog
+            mOAuthLoginDialog.show();
+          }
+        });
+
+        try {
+          // Wait for the OAuth browser page to be shown and user to login and for the CustomWebViewClient
+          // to get the auth code and request an OAuthTokenCredential with it
+          oAuthChallengeCountDownLatch.await();
+        } catch (InterruptedException ie) {
+        }
+        // Get the credential from the latch once it has been counted down
+        OAuthTokenCredential credential = oAuthChallengeCountDownLatch.getOAuthTokenCredential();
+        if (credential != null) {
+          ret = new AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CONTINUE_WITH_CREDENTIAL,
+              credential);
+        }
+        // dismiss the dialog
+        mOAuthLoginDialog.dismiss();
+      }
+      return ret != null ? ret : new AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CANCEL, challenge);
+    }
+  };
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_main);
+
+    // set custom challenge handler on the AuthenticationManager
+    AuthenticationManager.setAuthenticationChallengeHandler(mCustomAuthenticationChallengeHandler);
+
+    // initialize the dialog to display the OAuth login page
+    mOAuthLoginDialog = new Dialog(this);
+
+    // inflate MapView from layout
+    mMapView = (MapView) findViewById(R.id.mapView);
+
+    // add the OAuth configuration to the AuthenticationManager
+    try {
+      OAuthConfiguration config = new OAuthConfiguration(getString(R.string.portal_url), getString(R.string.client_id), getString(R.string.redirect_url));
+      AuthenticationManager.addOAuthConfiguration(config);
+    } catch (MalformedURLException e) {
+      throw new IllegalArgumentException(e.getMessage(), e.getCause());
+    }
+
+    // create a portal instance for the portal to load
+    Portal portal = new Portal(getString(R.string.portal_url));
+
+    // create a portal item with the itemId of the web map
+    PortalItem webMapItem = new PortalItem(portal, getString(R.string.webmap_id));
+
+    // create a map with the portal item
+    ArcGISMap map = new ArcGISMap(webMapItem);
+
+    // set the map to the map view
```

Comment:
worth to add here: "This will automatically initiate the loading of the map, which will cause the custom AuthenticationChallengeHandler to be invoked, which was previously set on the AuthenticationManager."

Reply:
added